modules/layers/src/contour-layer/marching-squares-codes.js

```diff
@@ -0,0 +1,444 @@
+const E_HEXAGON = [[-HALF, ONE6TH], [-HALF, -ONE6TH], OFFSET.S, OFFSET.SE, OFFSET.NE, OFFSET.N];
+const N_HEXAGON = [OFFSET.NW, OFFSET.W, [-ONE6TH, -HALF], [ONE6TH, -HALF], OFFSET.E, OFFSET.NE];
+const W_HEXAGON = [OFFSET.NW, OFFSET.SW, OFFSET.S, [HALF, -ONE6TH], [HALF, ONE6TH], OFFSET.N];
+const SW_NE_HEXAGON = [OFFSET.W, OFFSET.SW, OFFSET.S, OFFSET.E, OFFSET.NE, OFFSET.N];
+const NW_SE_HEXAGON = [OFFSET.NW, OFFSET.W, OFFSET.S, OFFSET.SE, OFFSET.E, OFFSET.N];
+
+// Heptagon (7-sided)
+const NE_HEPTAGON = [
+  [-HALF, ONE6TH],
+  [-HALF, -ONE6TH],
+  [-ONE6TH, -HALF],
+  [ONE6TH, -HALF],
+  OFFSET.E,
+  OFFSET.NE,
+  OFFSET.N
+];
+const SW_HEPTAGON = [
+  OFFSET.W,
+  OFFSET.SW,
+  OFFSET.S,
+  [HALF, -ONE6TH],
+  [HALF, ONE6TH],
+  [ONE6TH, HALF],
+  [-ONE6TH, HALF]
+];
+const NW_HEPTAGON = [
+  OFFSET.NW,
+  OFFSET.W,
+  [-ONE6TH, -HALF],
+  [ONE6TH, -HALF],
+  [HALF, -ONE6TH],
+  [HALF, ONE6TH],
+  OFFSET.N
+];
+const SE_HEPTAGON = [
+  [-HALF, ONE6TH],
+  [-HALF, -ONE6TH],
+  OFFSET.S,
+  OFFSET.SE,
+  OFFSET.E,
+  [ONE6TH, HALF],
+  [-ONE6TH, HALF]
+];
+
+// Octagon
+const OCTAGON = [
+  [-HALF, ONE6TH],
+  [-HALF, -ONE6TH],
+  [-ONE6TH, -HALF],
+  [ONE6TH, -HALF],
+  [HALF, -ONE6TH],
+  [HALF, ONE6TH],
+  [ONE6TH, HALF],
+  [-ONE6TH, HALF]
+];
+
+// Note: above wiki page invertes white/black dots for generating the code, we don't
+export const ISOLINES_CODE_OFFSET_MAP = {
+  // key is equal to the code of 4 vertices (invert the code specified in wiki)
+  // value can be an array or an Object
+  // Array : [line] or [line, line], where each line is [start-point, end-point], and each point is [x, y]
+  // Object : to handle saddle cases, whos output depends on mean value of all 4 corners
+  //  key: code of mean value (0 or 1)
+  //  value: Array , as above defines one or two line segments
+  0: [],
+  1: [[OFFSET.W, OFFSET.S]],
+  2: [[OFFSET.S, OFFSET.E]],
+  3: [[OFFSET.W, OFFSET.E]],
+  4: [[OFFSET.N, OFFSET.E]],
+  5: {
+    0: [[OFFSET.W, OFFSET.S], [OFFSET.N, OFFSET.E]],
+    1: [[OFFSET.W, OFFSET.N], [OFFSET.S, OFFSET.E]]
+  },
+  6: [[OFFSET.N, OFFSET.S]],
+  7: [[OFFSET.W, OFFSET.N]],
+  8: [[OFFSET.W, OFFSET.N]],
+  9: [[OFFSET.N, OFFSET.S]],
+  10: {
+    0: [[OFFSET.W, OFFSET.N], [OFFSET.S, OFFSET.E]],
+    1: [[OFFSET.W, OFFSET.S], [OFFSET.N, OFFSET.E]]
+  },
+  11: [[OFFSET.N, OFFSET.E]],
+  12: [[OFFSET.W, OFFSET.E]],
+  13: [[OFFSET.S, OFFSET.E]],
+  14: [[OFFSET.W, OFFSET.S]],
+  15: []
+};
+
+export const ISOBANDS_CODE_OFFSET_MAP = {
+  // Below list of cases, follow the same order as in above mentioned wiki page.
+  // Each case has its code on first commented line // T,TR,R,C
+  // where T: Top, TR: Top-right, R: Right and C: current, each will be either 0, 1 or 2
+  // final code is binary represenation of above code , where takes 2 digits
+  // for example:  code 2-2-2-1 => 10-10-10-01 => 10101001 => 169
+
+  // no contours
+
+  // 0000
+  0: [],
+  // 2222
+  170: [],
```

Comment:
Does it make sense to write it like this:

```js
[parseInt('2222', 4)]: []
```

To eliminate conversion errors?